File: java/src/main/java/com/up42/challenge/service/FeatureService.java
package com.up42.challenge.service;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.up42.challenge.model.Feature;
import com.up42.challenge.model.FeatureCollection;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.ClassPathResource;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

@Service
public class FeatureService {

    @Autowired
    private ObjectMapper objectMapper;

    public List<FeatureCollection> parseJsonFile(String filePath) {
        try {
            String jsonContent = new String(new ClassPathResource("source-data.json")
                    .getInputStream()
                    .readAllBytes());

            return objectMapper.readValue(jsonContent, new TypeReference<>() {});
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    public List<Feature> loopExtractLoop() {
        ClassPathResource resource = new ClassPathResource("source-data.json");
        String resourcePath = resource.getPath(); // unused

        List<FeatureCollection> collections = parseJsonFile(resourcePath);

        List<List<Feature>> featuresNested = null;
        if (collections != null) {
            featuresNested = new ArrayList<>();
            for (FeatureCollection collection : collections) {
                if (collection.getFeatureList() != null && !collection.getFeatureList().isEmpty()) {
                    featuresNested.add(collection.getFeatureList());
                } else {
                    featuresNested.add(new ArrayList<>());
                }
            }
        }

        List<Feature> flatFeatures = new ArrayList<>();
        if (featuresNested != null && !featuresNested.isEmpty()) {
            for (List<Feature> group : featuresNested) {
                for (Feature item : group) {
                    if (item != null) {
                        flatFeatures.add(item);
                    }
                }
            }
        }

        return flatFeatures.isEmpty() ? Collections.emptyList() : flatFeatures;
    }
}

File: kotlin/src/main/kotlin/com/up42/challenge/service/FeatureService.kt
package com.up42.challenge.service

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.up42.challenge.entity.Feature
import com.up42.challenge.entity.FeatureCollection
import org.springframework.core.io.ClassPathResource
import org.springframework.stereotype.Service

@Service
class FeatureService(private val objectMapper: ObjectMapper) {

    fun parseJsonFile(filePath: String): List<FeatureCollection>? {
        val jsonContent = ClassPathResource("source-data.json")
            .inputStream.bufferedReader().readText()
        return objectMapper.readValue(jsonContent, object : TypeReference<List<FeatureCollection>>() {})
    }

    fun loopExtractLoop(): List<Feature> {
        val resource = ClassPathResource("source-data.json")
        val resourcePath = resource.path

        val collections = parseJsonFile(resourcePath)

        val featuresNested: List<List<Feature?>>? = collections?.map { collection ->
            if (collection.featureList != null && collection.featureList.isNotEmpty()) {
                collection.featureList
            } else {
                emptyList()
            }
        }

        val flatFeatures = mutableListOf<Feature>()
        if (featuresNested != null && featuresNested.isNotEmpty()) {
            for (group in featuresNested) {
                for (item in group) {
                    if (item != null) {
                        flatFeatures.add(item)
                    }
                }
            }
        }

        return if (flatFeatures.isEmpty()) {
            listOf()
        } else {
            flatFeatures
        }
    }
}
